public/tracy/TracyMetal.hmm

```diff
@@ -0,0 +1,634 @@
+      and there's no way to reset them back to zero after timestamps get resolved; the only
+      way to clear the timestamps is by allocating a new timestamp query buffer
+   c) if a command encoder records no commands and its corresponding command buffer ends up
+      committed to the command queue, Metal will "optimize-away" the encoder along with any
+      timestamp queries associated with it (the timestamp will remain as zero and will never
+      get resolved)
+   Because of the limitations above, two timestamp buffers are managed internally. Once one
+   of the buffers fills up with requests, the second buffer can start serving new requests.
+   Once all requests in a buffer get resolved and collected, the entire buffer is discarded
+   and a new one allocated for future requests. (Proper cycling through a ring buffer would
+   require bookkeeping and completion handlers to collect only the known complete queries.)
+   In the current implementation, there is potential for a race condition when the buffer is
+   discarded and reallocated. In practice, the race condition will never materialize so long
+   as TracyMetalCollect() is called frequently to keep the amount of unresolved queries low.
+   Finally, there's a timeout mechanism during timestamp collection to detect "empty" command
+   encoders and ensure progress.
+*/
+
+#ifndef TRACY_ENABLE
+
+#define TracyMetalContext(device) nullptr
+#define TracyMetalDestroy(ctx)
+#define TracyMetalContextName(ctx, name, size)
+
+#define TracyMetalZone(ctx, name)
+#define TracyMetalZoneC(ctx, name, color)
+#define TracyMetalNamedZone(ctx, varname, name, active)
+#define TracyMetalNamedZoneC(ctx, varname, name, color, active)
+#define TracyMetalZoneTransient(ctx, varname, name, active)
+
+#define TracyMetalZoneS(ctx, name, depth)
+#define TracyMetalZoneCS(ctx, name, color, depth)
+#define TracyMetalNamedZoneS(ctx, varname, name, depth, active)
+#define TracyMetalNamedZoneCS(ctx, varname, name, color, depth, active)
+#define TracyMetalZoneTransientS(ctx, varname, name, depth, active)
+
+#define TracyMetalCollect(ctx)
+
+namespace tracy
+{
+class MetalZoneScope {};
+}
+
+using TracyMetalCtx = void*;
+
+#else
+
+#include <atomic>
+#include <assert.h>
+#include <stdlib.h>
+
+#include "Tracy.hpp"
+#include "../client/TracyProfiler.hpp"
+#include "../client/TracyCallstack.hpp"
+#include "../common/TracyAlign.hpp"
+#include "../common/TracyAlloc.hpp"
+
+// ok to import if in obj-c code
+#import <Metal/Metal.h>
+
+#define VA_ARGS(...) , ##__VA_ARGS__
+
+#define TracyMetalPanic(ret, msg, ...) do { \
+    char buffer [1024]; \
+    snprintf(buffer, sizeof(buffer), "TracyMetal: " msg VA_ARGS(__VA_ARGS__)); \
+    TracyMessageC(buffer, strlen(buffer), tracy::Color::OrangeRed); \
+    fprintf(stderr, "%s\n", buffer); \
+    assert(false && "TracyMetal: " msg); \
+    ret; \
+    } while(false);
+
+
+#define TRACY_METAL_DEBUG_MASK (0)
```

Comment:
Yes, at this initial stage, I'd like to have the verbose debugging log available when needed.
Should someone report issues, I can ask them to set this define prior to including the header.

Reply:
It is understood that you want people to possibly override the macro, but the right coding pattern to do so is to wrap the `#define` with `#ifndef ... #endif` which is missing as of writing to avoid compiler warnings (and thus maybe errors) in client codebase.

Reply:
Ah, LOL, yeah, I did not realize this was a code change _suggestion_. You are absolutely right!

Reply:
done!